### PR TITLE
Make javac annotation-processor-related options configurable in CompilerOptions

### PR DESCRIPTION
**Pull request by tbroyer**
-processorpath, -processor and -Akey=value processor arguments.

See discussion at https://groups.google.com/d/topic/gradle-dev/w3aQgjY7mQk

### CODE REVIEW

**Comment by breskeby** [comment]
This PR is related to #456. One thing that you should have in place here is reasonable integration test coverage. Especially for ensuring we don't introduce any breaking changes.

**Comment by breskeby** [comment]
Hey @tbroyer,

any update on this?

**Comment by tbroyer** [comment]
Was working on it. PTAL?

BTW, I'll have a few amendments to make to the design-doc: remove `proc` (`proc:only` won't play well with incremental builds as it won't produce anything in `destinationDir`), and add a `generatedSourcesDir` to control javac's `-s` option.
Would you prefer a separate PR for the design doc than for the implementation (not done yet anyway) or is it OK including both in the same PR? (same commit?) And in any case, should I update this PR with `generatedSourcesDir` (when I implement it) or would you prefer a separate PR?

**Comment by breskeby** [comment]
sorry for the late reply. I'd prefere to keep the design doc separated for now. thanks

**Comment by tbroyer** [comment]
See #616 for the design-doc update

File: subprojects/plugins/src/integTest/groovy/org/gradle/java/compile/BasicJavaCompilerIntegrationSpec.groovy [hunk 1]
```diff
@@ -252,91 +252,171 @@ class Main {
...
+        writeAnnotationProcessorProjects("annotation", "processor", "SimpleAnnotation", "Generated")
+
+        file("src/main/java/Java.java") << "@com.test.SimpleAnnotation public class Java {}"
+
+        then:
+        succeeds("compileJava")
+        file('build/classes/main/Java$$Generated.java').exists()
+    }
+
+    @LeaksFileHandles("holds processor.jar open for in process compiler")
```
Comment (breskeby) [inline subprojects/plugins/src/integTest/groovy/org/gradle/java/compile/BasicJavaCompilerIntegrationSpec.groovy]: Can you elaborate why this is the case? Why is the processor.jar file reference not closed after the test? Just to be clear, the `LeakFileHandles` annotation is just a marker interface for our integration tests to document where our test infrastructure file handling is not optimal

File: subprojects/language-java/src/test/groovy/org/gradle/api/internal/tasks/compile/JavaCompilerArgumentsBuilderTest.groovy [hunk 2]
```diff
@@ -327,6 +327,29 @@ class JavaCompilerArgumentsBuilderTest extends Specification {
...
+    }
+
+    def "generates -processor option"() {
+        spec.compileOptions.processors = ["foo.FooProcessor", "bar.BarProcessor"]
+
+        expect:
+        builder.build() == ["-g", "-processor", "foo.FooProcessor,bar.BarProcessor", USE_UNSHARED_COMPILER_TABLE_OPTION]
+    }
+
+    def "generates processor arguments options"() {
```
Comment (breskeby) [inline subprojects/language-java/src/test/groovy/org/gradle/api/internal/tasks/compile/JavaCompilerArgumentsBuilderTest.groovy]: I'd name it "can declare processor arguments"

File: subprojects/plugins/src/integTest/groovy/org/gradle/java/compile/BasicJavaCompilerIntegrationSpec.groovy [hunk 3]
```diff
@@ -252,91 +252,171 @@ class Main {
...
+        file("src/main/java/Java.java") << "@com.test.SimpleAnnotation1 @com.test.SimpleAnnotation2 public class Java {}"
+
+        then:
+        succeeds("compileJava")
+        file('build/classes/main/Java$$Generated1.java').exists()
+        !file('build/classes/main/Java$$Generated2.java').exists()
+    }
+
+    def writeAnnotationProcessorProjects(String annotationProject, String processorProject, String annotationName, String generatedClassSuffix) {
+        file(annotationProject).create {
```
Comment (breskeby) [inline subprojects/plugins/src/integTest/groovy/org/gradle/java/compile/BasicJavaCompilerIntegrationSpec.groovy]: To keep the test classes small and readable you should prefer to use the `TestResources` JUnit Rule to organize test project code. Check the JUnitIntegrationTest as an example how to use this Rule.

**Comment by bmuschko** [comment]
@tbroyer Are you planning to make the proposed changes to the pull request soon?

**Comment by tbroyer** [comment]
Will revisit ASAP; should I include the changes described in #616 (`annotationProcessing`, possibly also `generatedSourcesDestinationDir`) or would you prefer a separate PR? (given #616 hasn't been merged)

File: subprojects/plugins/src/integTest/groovy/org/gradle/java/compile/BasicJavaCompilerIntegrationSpec.groovy (hunk 3 above)
Comment (tbroyer) [inline subprojects/plugins/src/integTest/groovy/org/gradle/java/compile/BasicJavaCompilerIntegrationSpec.groovy]: That's no longer possible (or not that easy), now that the processor's template string makes use of a method of the test class (`gradleLeaksIntoAnnotationProcessor`) that's overridden in subclasses. That would mean creating 4 TestResources "projects" (one with annotation+processor modules assuming Gradle doesn't leak into annotation processors, the same but that time assuming Gradle does not leak, and a similar pair with annotation+processor modules with other names)

File: subprojects/plugins/src/integTest/groovy/org/gradle/java/compile/BasicJavaCompilerIntegrationSpec.groovy (hunk 1 above)
Comment (tbroyer) [inline subprojects/plugins/src/integTest/groovy/org/gradle/java/compile/BasicJavaCompilerIntegrationSpec.groovy]: I have no idea, this is a "cargo-cult copy-paste" from the existing `"can use annotation processor"` test.

**Comment by tbroyer** [comment]
Updated. For some reason, `:plugins:integTests` seem to deadlock (running with `--debug`, it seems to come from a Java serialization issue talking to the daemon in `DaemonJavaCompilerIntegrationTest`; I don't know enough about the daemon to debug this, any hint appreciated).

**Comment by eriwen** [comment]
Hi @tbroyer, I'm having a hard time reproducing the deadlock you're reporting running `:plugins:integTest` in debug mode. If you would provide steps to reproduce, I'm happy to advise.

**Comment by tbroyer** [comment]
```bash
$ uname -a
Linux M2800-2 4.4.0-31-generic #50-Ubuntu SMP Wed Jul 13 00:07:12 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
$ java -version
java version "1.8.0_101"
Java(TM) SE Runtime Environment (build 1.8.0_101-b13)
Java HotSpot(TM) 64-Bit Server VM (build 25.101-b13, mixed mode)

$ ./gradlew clean :plugins:integTest
…
> Building 99% > :plugins:integTest > 48 tests completed, 6 skipped
```

If I run with `--debug`, the log ends up with an error, then fills up with the 4 last lines:
```bash
$ ./gradlew clean :plugins:integTest --debug
…
17:56:55.497 [DEBUG] [TestEventLogger]     Starting process 'Gradle Compiler Daemon 1'. Working directory: /mnt/ssd/tbr/Projets/gradle/subprojects/plugins/build/tmp/test files/DaemonJavaCompilerIntegrationTest/can_use_annotation_...essorpath/z5rfy Command: /usr/lib/jvm/java-8-oracle/bin/java -Djava.security.manager=worker.org.gradle.process.internal.worker.child.BootstrapSecurityManager -Dfile.encoding=UTF-8 -Duser.country=FR -Duser.language=fr -Duser.variant -cp /mnt/ssd/tbr/Projets/gradle/intTestHomeDir/worker-1/caches/3.0-snapshot-1/workerMain/gradle-worker.jar worker.org.gradle.process.internal.worker.GradleWorkerMain 'Gradle Compiler Daemon 1'
17:56:55.497 [DEBUG] [TestEventLogger] 
17:56:55.498 [DEBUG] [TestEventLogger]     Successfully started process 'Gradle Compiler Daemon 1'
17:56:55.498 [DEBUG] [TestEventLogger] 
17:56:55.692 [DEBUG] [TestEventLogger]     Started Gradle compiler daemon (0.236 secs) with fork options DaemonForkOptions{minHeapSize=null, maxHeapSize=null, jvmArgs=[], classpath=[]}.
17:56:55.693 [DEBUG] [TestEventLogger] 
17:56:55.697 [DEBUG] [TestEventLogger] 
17:56:55.697 [DEBUG] [TestEventLogger] org.gradle.java.compile.daemon.DaemonJavaCompilerIntegrationTest > can use annotation processor in processorpath STANDARD_ERROR
17:56:55.697 [DEBUG] [TestEventLogger]     Unexpected exception thrown.
17:56:55.697 [DEBUG] [TestEventLogger] 
17:56:55.697 [DEBUG] [TestEventLogger]     org.gradle.internal.remote.internal.MessageIOException: Could not write message [ChannelMessage channel: org.gradle.process.internal.worker.request.RequestProtocol, payload: [MethodInvocation method: run(execute, [Ljava.lang.Class;@18274eea, [Ljava.lang.Object;@1c18d936)]] to '/0:0:0:0:0:0:0:1:42760'.
17:56:55.697 [DEBUG] [TestEventLogger] 
17:56:55.697 [DEBUG] [TestEventLogger]     	at org.gradle.internal.remote.internal.inet.SocketConnection.dispatch(SocketConnection.java:111)
17:56:55.697 [DEBUG] [TestEventLogger] 
17:56:55.697 [DEBUG] [TestEventLogger]     	at org.gradle.internal.remote.internal.hub.MessageHub$ConnectionDispatch.run(MessageHub.java:298)
17:56:55.697 [DEBUG] [TestEventLogger] 
17:56:55.697 [DEBUG] [TestEventLogger]     	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:54)
17:56:55.697 [DEBUG] [TestEventLogger] 
17:56:55.697 [DEBUG] [TestEventLogger]     	at org.gradle.internal.concurrent.StoppableExecutorImpl$1.run(StoppableExecutorImpl.java:40)
17:56:55.697 [DEBUG] [TestEventLogger] 
17:56:55.697 [DEBUG] [TestEventLogger]     	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
17:56:55.697 [DEBUG] [TestEventLogger] 
17:56:55.697 [DEBUG] [TestEventLogger]     	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
17:56:55.697 [DEBUG] [TestEventLogger] 
17:56:55.697 [DEBUG] [TestEventLogger]     	at java.lang.Thread.run(Thread.java:745)
17:56:55.697 [DEBUG] [TestEventLogger] 
17:56:55.697 [DEBUG] [TestEventLogger]     Caused by: java.io.NotSerializableException: org.gradle.api.internal.artifacts.configurations.DefaultConfiguration_Decorated
17:56:55.697 [DEBUG] [TestEventLogger] 
17:56:55.697 [DEBUG] [TestEventLogger]     	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1184)
17:56:55.697 [DEBUG] [TestEventLogger] 
17:56:55.697 [DEBUG] [TestEventLogger]     	at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1548)
17:56:55.698 [DEBUG] [TestEventLogger] 
17:56:55.698 [DEBUG] [TestEventLogger]     	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1509)
17:56:55.698 [DEBUG] [TestEventLogger] 
17:56:55.698 [DEBUG] [TestEventLogger]     	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1432)
17:56:55.698 [DEBUG] [TestEventLogger] 
17:56:55.698 [DEBUG] [TestEventLogger]     	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1178)
17:56:55.698 [DEBUG] [TestEventLogger] 
17:56:55.698 [DEBUG] [TestEventLogger]     	at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1548)
17:56:55.698 [DEBUG] [TestEventLogger] 
17:56:55.698 [DEBUG] [TestEventLogger]     	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1509)
17:56:55.698 [DEBUG] [TestEventLogger] 
17:56:55.698 [DEBUG] [TestEventLogger]     	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1432)
17:56:55.698 [DEBUG] [TestEventLogger] 
17:56:55.698 [DEBUG] [TestEventLogger]     	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1178)
17:56:55.698 [DEBUG] [TestEventLogger] 
17:56:55.698 [DEBUG] [TestEventLogger]     	at java.io.ObjectOutputStream.writeArray(ObjectOutputStream.java:1378)
17:56:55.698 [DEBUG] [TestEventLogger] 
17:56:55.698 [DEBUG] [TestEventLogger]     	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1174)
17:56:55.698 [DEBUG] [TestEventLogger] 
17:56:55.698 [DEBUG] [TestEventLogger]     	at java.io.ObjectOutputStream.writeArray(ObjectOutputStream.java:1378)
17:56:55.698 [DEBUG] [TestEventLogger] 
17:56:55.699 [DEBUG] [TestEventLogger]     	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1174)
17:56:55.699 [DEBUG] [TestEventLogger] 
17:56:55.699 [DEBUG] [TestEventLogger]     	at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:348)
17:56:55.699 [DEBUG] [TestEventLogger] 
17:56:55.699 [DEBUG] [TestEventLogger]     	at org.gradle.internal.serialize.Message.send(Message.java:40)
17:56:55.699 [DEBUG] [TestEventLogger] 
17:56:55.699 [DEBUG] [TestEventLogger]     	at org.gradle.internal.remote.internal.hub.JavaSerializationBackedMethodArgsSerializer$ArraySerializer.write(JavaSerializationBackedMethodArgsSerializer.java:59)
17:56:55.699 [DEBUG] [TestEventLogger] 
17:56:55.699 [DEBUG] [TestEventLogger]     	at org.gradle.internal.remote.internal.hub.JavaSerializationBackedMethodArgsSerializer$ArraySerializer.write(JavaSerializationBackedMethodArgsSerializer.java:51)
17:56:55.699 [DEBUG] [TestEventLogger] 
17:56:55.699 [DEBUG] [TestEventLogger]     	at org.gradle.internal.remote.internal.hub.MethodInvocationSerializer$MethodInvocationWriter.writeArgs(MethodInvocationSerializer.java:75)
17:56:55.699 [DEBUG] [TestEventLogger] 
17:56:55.699 [DEBUG] [TestEventLogger]     	at org.gradle.internal.remote.internal.hub.MethodInvocationSerializer$MethodInvocationWriter.write(MethodInvocationSerializer.java:71)
17:56:55.699 [DEBUG] [TestEventLogger] 
17:56:55.699 [DEBUG] [TestEventLogger]     	at org.gradle.internal.remote.internal.hub.MethodInvocationSerializer$MethodInvocationWriter.write(MethodInvocationSerializer.java:56)
17:56:55.699 [DEBUG] [TestEventLogger] 
17:56:55.699 [DEBUG] [TestEventLogger]     	at org.gradle.internal.serialize.kryo.TypeSafeSerializer$2.write(TypeSafeSerializer.java:43)
17:56:55.699 [DEBUG] [TestEventLogger] 
17:56:55.699 [DEBUG] [TestEventLogger]     	at org.gradle.internal.remote.internal.hub.InterHubMessageSerializer$MessageWriter.write(InterHubMessageSerializer.java:102)
17:56:55.699 [DEBUG] [TestEventLogger] 
17:56:55.699 [DEBUG] [TestEventLogger]     	at org.gradle.internal.remote.internal.hub.InterHubMessageSerializer$MessageWriter.write(InterHubMessageSerializer.java:87)
17:56:55.699 [DEBUG] [TestEventLogger] 
17:56:55.699 [DEBUG] [TestEventLogger]     	at org.gradle.internal.remote.internal.inet.SocketConnection.dispatch(SocketConnection.java:109)
17:56:55.699 [DEBUG] [TestEventLogger] 
17:56:55.699 [DEBUG] [TestEventLogger]     	... 6 more
17:56:55.700 [DEBUG] [TestEventLogger] 
17:56:55.774 [DEBUG] [TestEventLogger] 
17:56:55.774 [DEBUG] [TestEventLogger] org.gradle.java.compile.daemon.DaemonJavaCompilerIntegrationTest > can use annotation processor in processorpath STANDARD_OUT
17:56:55.774 [DEBUG] [TestEventLogger]     Initialized native services in: /mnt/ssd/tbr/Projets/gradle/intTestHomeDir/worker-1/native
17:56:55.775 [DEBUG] [TestEventLogger] 
17:56:56.423 [DEBUG] [org.gradle.launcher.daemon.server.Daemon] DaemonExpirationPeriodicCheck running
17:56:56.423 [DEBUG] [org.gradle.cache.internal.DefaultFileLockManager] Waiting to acquire shared lock on daemon addresses registry.
17:56:56.423 [DEBUG] [org.gradle.cache.internal.DefaultFileLockManager] Lock acquired.
17:56:56.424 [DEBUG] [org.gradle.cache.internal.DefaultFileLockManager] Releasing lock on daemon addresses registry.
17:57:06.423 [DEBUG] [org.gradle.launcher.daemon.server.Daemon] DaemonExpirationPeriodicCheck running
17:57:06.423 [DEBUG] [org.gradle.cache.internal.DefaultFileLockManager] Waiting to acquire shared lock on daemon addresses registry.
17:57:06.423 [DEBUG] [org.gradle.cache.internal.DefaultFileLockManager] Lock acquired.
17:57:06.424 [DEBUG] [org.gradle.cache.internal.DefaultFileLockManager] Releasing lock on daemon addresses registry.
17:57:16.423 [DEBUG] [org.gradle.launcher.daemon.server.Daemon] DaemonExpirationPeriodicCheck running
17:57:16.423 [DEBUG] [org.gradle.cache.internal.DefaultFileLockManager] Waiting to acquire shared lock on daemon addresses registry.
17:57:16.423 [DEBUG] [org.gradle.cache.internal.DefaultFileLockManager] Lock acquired.
17:57:16.424 [DEBUG] [org.gradle.cache.internal.DefaultFileLockManager] Releasing lock on daemon addresses registry.
```

So, maybe not a deadlock per se, but still the process never finishes unless I terminate it.

**Comment by bmuschko** [comment]
@eriwen Did you have a chance to follow up based on the information provided by @tbroyer?

**Comment by tbroyer** [comment]
Gradle 3.4 introduced `annotationProcessorPath`, and I'm not so sure anymore `-processor` and `-Akey=value` are worth it.

**Pull request closed without merging**